### ISSUE TITLE
Minimum viable package manager

### ISSUE THREAD

**Issue by laurmaedje** [typst/typst#1554]
Building a fully-fledged package manager is complex and too big of a time investment currently. Nonetheless, package management is one of the most asked-for features. That's why we have decided to build a **minimum viable package manager** while leaving options for expansion open for the future.

## Technical design proposal
- Packages live on https://github.com/typst/packages
- Packages are submitted through pull requests (at least for now)
- Each package version is a separate folder in the repository
- Packages must have a TOML file with metadata, license, etc.
- A GitHub action builds tarballs for all packages and uploads them to Azure blob storage
- Package imports always need to specify the exact version (at least for now). This way, we don't need to deal with locking or anything like that. While this more easily leads to situations where a package is included in multiple versions, that's not really a big problem.
- When a package is imported / included in the CLI, it searches for it in a well-known system location. If not found, it downloads it from its well-known URL. Because packages are immutable and full versions must be specified, the network is not touched if all packages are already available.
- Similarly, when a package is imported / included in the web app, it searches for it / stores it in the IndexDB
- A searchable list of packages with documentation will be available on https://typst.app/docs/packages. For now, packages can have one Markdown file of documentation (no Typst HTML export yet :/). In the future, we can have automatic documentation generation à la rustdoc. 
- Search index and docs are also built via GitHub action. While this doesn't scale, it should take us far enough (the process for each individual package is relatively cheap).

We are not yet decided on namespacing. Namespaces could be very useful, for instance for organization or university templates. A global namespace on the other hand is a lot more discoverable for packages. One possibility that would allow us to keep all options open, is to have _all_ packages in a `@preview` namespace for now (e.g. `#import "@preview/somepackage:0.2"`).

## Submissions
There will be guidelines for what criteria a package needs to fulfill to be published. These will be reviewed upon submission via pull request. Below is a first draft of the submission guidelines. These guidelines are up for discussion!

- Naming
  - Names should not include the word `typst` as it is redundant
  - Names should not be merely descriptive (e.g. not just `slides`) to create level grounds for everybody, without one package immediately being the "blessed" solution
  - Names (and all package contents) must be safe for work
  Currently, there are many projects called `typst-thing`. To be published as packages, these will need to come up with a more creative name. Merely descriptive names _could_ maybe be used in the future if things do become the blessed solution / an extension to the standard library

- Documentation: All package exports intended for public use must be (at least briefly) documented in the accompanying Markdown file

- Size: Packages should not contain large files or a large number of files. This will be judged on a case-by-case basis, but if it needs more than ten files, it should be well-motivated.

- License: Packages must be licensed under the terms of an OSI-approved license

## Open Questions
- Best system location for package storage:
  - `${XDG_CACHE_DIR}/typst/packages/*` on Unix?
  - Somewhere in `AppData` on Windows?
- Should we go with the `@preview` namespace?
- Any changes or additions to the submission guidelines?

Any feedback on the design would be welcome. We plan to build this package manager in a matter of days and ship it in the next release.

**Comment by alerque** [comment]
Having experience with package managers (packager for multiple Linux distros, author of typesetting software with external packages, maintainer of vim plugins installable via multiple plugin managers, etc.) I would suggest this approach is going to be *very* hard to maintain long term. It isn't a terrible idea to get the ball rolling but you'll start having trouble with it when things start accruing.

I would highly recommend you go with the "everything is namespaced `@preview`" approach to make it much easier to migrate to next iteration of packages once you outgrow this approach.

**Comment by edhebi** [comment]
A few thoughts:

#### names and namespacing

If we want to have distinct names as to avoid "blessed ones", I feel that we should have namespacing. If I currently have `typst-thingy` and I can't/don't want to give it a fancy name, I'd much rather call it `@edhebi/thingy` than `edhebi-thingy`. This is also useful in a more direct manner because for a direct import it makes it available in a `thingy` module rather than a `edhebi-thingy` module, which for long package names feel like it would be an issue fairly quicly

All that being said, I do think a `@preview` namespace is the right call in the beginning, maybe even `@preview/v1`.

#### Paths

Please do not default to a dot directory right under the home and instead follow the [XDG Base Directories](https://wiki.archlinux.org/title/XDG_Base_Directory) spec. This is a defacto standard because a lot of people are realy tired of home directories pollution. In the case of typst, the result would be storing packages in `${XDG_CACHE_DIR}/typst/packages/*` and defaulting to `$HOME/.cache/typst/packages/*`. This also works on other OSes just by using a different fallback values for that directory.

#### Immutability

There's nothing technically preventing a package from being changed on the website, including an existing version of a package. I'd like to see hashing of packages at some point of the process. having to provide those along with the version would be incredibly unwieldy tho

**Comment by sascha-andres** [comment]
Reg immutability: having a vendor folder would solve this on a project level

Hashing would be sth like a lock file and checks against this

**Comment by astrale-sharp** [comment]
@preview namespace is definitely the way to go for a temporary solution like this.

is there a risk with packages (can we still read ssh and invisibly embed it in a pdf)? 

if not another solution would be to have a `#import "@github.com/astrale-sharp/my_super_pck` and maybe an option to specify the hash.
checks at runtime if the folder is conveniently organised ( helped by a typst-package.toml file )

ultimately it's free and people understand that it's the responsability of the owner of the package.

This could live alongside @preview where you have the assurance things have been reviewed and won't change.

Some might argue it's a terrible idea but I'm not so sure

**Comment by Dherse** [comment]
> is there a risk with packages (can we still read ssh and invisibly embed it in a pdf)?
I believe that this falls under another PR/RFC that has been opened that discusses a basic permission systems as a temporary stopgap.

**Comment by PgBiel** [comment]
Following @alerque and @astrale-sharp, while I can see the benefits of hashing out a package manager quickly, perhaps a better idea for now might be to allow users to import packages directly from GitHub à la Deno. Something like `#import "@git/github.com/PgBiel/typst-tablex.git"` which would take the latest commit from that repository, while adding some specifier like a #commithash or a #v0.0.3 tag would allow being more specific. (Note the @git prefix, which would be generic enough to allow for other sites such as GitLab, codeberg etc.)

I feel this would be better as a temporary solution, as "there is nothing more permanent than a temporary solution" as they say, and this git approach would be permanently viable, while a unified repository requires a _lot_ of maintenance, which I'm not sure would be the best thing for the Typst team to focus on.
Perhaps it would be nicer to keep that for a more ideal future where Typst has its own package distribution solution instead, which would likely also allow for more flexibility in general.

Anyways, just my two cents :)

**Comment by laurmaedje** [comment]
What I don't like about the git approach is that:

- There's no central place to discover packages: No search, no docs, no autocomplete for packages
- There is no guarantee that some package doesn't randomly change or disappear (even if you add a hash, which realistically nobody will do, it can still disappear)
- We have to deal with whatever people put into their repositories. The central repository approach allows us to discover potential problems through the PR process and adapt the submission guidelines accordingly. I'm optimistic that the available packages will be easier and safer to use and have better documentation.
- To access packages in the web app, it would either need to do cross-origin requests to GitHub (difficult becomes of COEP, could download tons of data) or allow random GitHub downloads through our backend (not a fan)
- More difficult to use packages for templates in the template gallery. With the central approach that can simply be configured through a key in the manifest file

In my opinion, the central approach allows for a simpler, safer, and more guided user experience. And good user experience is what Typst is about.

**Comment by PgBiel** [comment]
I agree a central repository is better in general, and would solve the problems you mention, but will the Typst team be able (/ have the time) to maintain it properly? Will it scale properly, when the repository starts inevitably accumulating tons of pull requests (as Typst's popularity increases)? That's my main concern, following @alerque's comments.

I'm not entirely sure how to best tackle this right now. However, if the proposal is implemented as is, we must already have in mind a migration plan for when Typst's own package distribution solution (something like crates.io) is inevitably rolled out (which would likely make everything a lot easier, once the initial heavy work is done). Of course, such a change, if it occurs, would come only in the far future, though, as I'm sure it would require having its own team of people working on it all the time (preferably separate from the compiler team).

**Comment by laurmaedje** [comment]
Packages would be distributed through a REST endpoint. Thus, a migration to a crates.io-like registry would only change the package submission process, not the package retrieval process. It wouldn't break anything, we could simply push all existing packages from GitHub to that new registry and serve the REST API from there. The repository approach obviously won't scale forever, but it can bring a good user experience now (even if the package author experience is not yet ideal).

**Comment by clbarnes** [comment]
How about a package monorepo where each submission is a submodule? This has a number of advantages:

- It isn't unwieldy to clone the repo as an index - then the tooling only has to check out the requested packages
- Development happens in different repositories, so developers get to use their preferred host and don't need every other package in order to develop their own
- PRs updating submissions have minimal git churn

I think submodules' git histories are fully navigable?

**Comment by sascha-andres** [comment]
Regarding packages going away: how to prevent this anyways? As a package author I decide on license and stuff. The chances are high, that there might be packages that would disallow feeding a package manager source from a different golden source.

Or the property and rights on packages and code  would have to be transferred to some foundation or so.

As much as I would like to see stable packages, it might even be an obligation to remove a package (eg a court ruling).

**Comment by reknih** [comment]
> Regarding packages going away: how to prevent this anyways? As a package author I decide on license and stuff.

We will exclusively accept packages with OSI-approved licenses so some sanity ground rules are ensured.

> As much as I would like to see stable packages, it might even be an obligation to remove a package (eg a court ruling).

We will not be able to guarantee a package will never be removed. (Could violate any local jurisdictions rules, be harmful to someone, or otherwise in breach of our Terms of Service...) This isn't a block chain. But under normal circumstances, any given version of a package will not be changed or removed once published. Think about it as a convention.

**Comment by alerque** [comment]
Actually removal could be important. Not for example crates.io also guarantees that published packages won't change, but it does have a yank feature for removing problematic crates.

**Comment by Dherse** [comment]
> a yank feature for removing problematic crates.

While I can see how this is useful for rust, I fail to see in which circumstances that could be useful for packages in typst, after all, typst isn't expected (hopefully) to make web queries, etc., therefore, security issues or breaking bugs should not be expected. And for breaking bugs, I'd rather see a warning in the CLI warning me that the version is "not recommended for use" or something similar than see my document break.

I think this is of particular interest in scenarios where you're using automated documentation building pipeline: the pipeline might be five years old, you still expect it to work as the day you set it up.

**Comment by sascha-andres** [comment]
> While I can see how this is useful for rust, I fail to see in which circumstances that could be useful for packages in typst, after all, typst isn't expected (hopefully) to make web queries, etc., therefore, security issues or breaking bugs should not be expected.

While this is true currently this does not resolve issues with sth like malicious images. This would not be a problem with typst itself but rather libraries and a package author including content which utilizes the exploit. I think you want to pull such a package. Not to talk about non technical issues.

> I think this is of particular interest in scenarios where you're using automated documentation building pipeline: the pipeline might be five years old, you still expect it to work as the day you set it up.

Agreed. But I also want to trust my build not to have unintended side effect. Not now, nor in five years time. The authors of the tooling I use are not responsible nor have they intended a malicious use case. Unfortunately, in the past, embedded images were an attack vector. And such content would be transportable using packages. Though I am definitely not some who can judge on the real risk of something like this.

**Comment by laurmaedje** [comment]
The Typst compiler now has package management support on main. You can install packages locally or from the global package repository. The repository and all further details are on https://github.com/typst/packages. 

We'd be glad if you'd try creating some packages and experimenting with them locally. If you find any issues, feel free to ping me. We will start accepting submissions to the shared package repository tomorrow. We'll be sure to create a PR template before that.

If everything works out fine, a new release and web app version with package management support will ship within a few days.
